Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,37 +1,345 @@
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <stdbool.h>
 
 #define SUCCESS 0
 #define ERROR -1
-#define OVERFLOW -2
-#define EMPTY -3
+// Defined limit for a string key length 
+#define LIMIT 10
 
-// Hash struct
-typedef struct Hash {
-    int key;
+// Hash Item struct
+typedef struct Item {
+    char* key;
     int value;
-} Hash;
+} Item;
 
-void print() {
+// Hash Table struct
+typedef struct HashTable {
+    int size;
+    Item** items;
+} HashTable;
 
+
+// returns true if a collision would happen for the initially hashed index for a key/value pair and false if no collision would occurred 
+bool isCollision(HashTable * table, int index) {
+    if (table->items[index] != NULL) {
+        return true;
+    }
+    return false;
 }
 
-void traverse() {
+// equals checks the testHash and expectedHash for parity
+// if the two HashTables match returns true
+// if a mismatch occurs returns false
+bool equals(HashTable * testHash, HashTable * expectedHash) {
+    for (int i = 0; i < expectedHash->size; i++) {
+        if (strcmp(testHash->items[i]->key, expectedHash->items[i]->key) != 0 || testHash->items[i]->value != expectedHash->items[i]->value) {
+            return false;
+        }
+    }
+    return true;
+}
 
+// hashing 
+// returns a "hashed" index based on modulus of the key's ASCII integer sum and the SIZE of the hashTable
+// Explanation:  I used loselose because it was a simple hashing function that seemed adequate for the purposes of this exercise.
+int hashing(HashTable *hashtable, char *key ) {
+    int i = 0;
+    int count = 0;
+    while (key[i] != NULL) {
+        count += key[i];
+        i++;
+    }
+    return count % hashtable->size;
 }
 
-bool compare() {
-    return true;
-    return false;
+
+// creates a new Item struct
+// returns new struct
+Item * createItem(char key[], int value, int * success) {
+    Item * item = malloc(sizeof(Item));
+    if (item != NULL) {        
+        item->key = strdup(key);
+        item->value = value;
+        *success = SUCCESS;
+        return item;
+    }
+    return NULL;
 }
 
+// creates a new HashTable struct
+// returns new HashTable struct
+HashTable * createHashTable(int size) {
+    if (size < 1 || size > 100) return NULL;
```

Comment:
Is there a reason why > 100 returns NULL?

Reply:
I wanted a check for size limits  but can't return an integer to signify ERROR like normal since the function expects a struct to be returned.

Reply:
Is there any reason why 100 is max?

Reply:
I thought it would be a reasonable limit for a hashTable - Do you think there shouldn't be a limit or maybe I am just going too large for scope?

Reply:
Unless there is an algorithmic constraint that makes the implementation unusable, I would not put any upper limit at all - we can assume that clients can do and should do their own memory planning.

Reply:
Understood! Removed the line to check for an upper limit: [commit 67427531039df54c6437a39b4ff65817a40e4100](https://github.com/kcarrel/Training-Notes/pull/12/commits/67427531039df54c6437a39b4ff65817a40e4100)